diff --git a/CMakeLists.txt b/CMakeLists.txt
new file mode 100644
index 0000000..9635b8c
--- /dev/null
+++ b/CMakeLists.txt
@@ -0,0 +1,17 @@
+add_library(libfmq SHARED
+        EventFlag.cpp
+        FmqInternal.cpp
+)
+
+target_include_directories(libfmq PUBLIC include)
+target_link_libraries(libfmq libutils libcutils libbase)
+
+set_target_properties(libfmq PROPERTIES
+    VERSION ${PROJECT_VERSION_MAJOR}.${PROJECT_VERSION_MINOR}.${PROJECT_VERSION_PATCH}
+)
+
+install(TARGETS libfmq
+        LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+        PUBLIC_HEADER)
+
+install(DIRECTORY include/fmq DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/linux-hidl)
