diff --git a/src/libhwbinder/CMakeLists.txt b/src/libhwbinder/CMakeLists.txt
new file mode 100644
index 0000000..96e60af
--- /dev/null
+++ b/src/libhwbinder/CMakeLists.txt
@@ -0,0 +1,25 @@
+add_library(libhwbinder SHARED
+	Binder.cpp
+	BpHwBinder.cpp
+	BufferedTextOutput.cpp
+	Debug.cpp
+	IInterface.cpp
+	IPCThreadState.cpp
+	Parcel.cpp
+	ProcessState.cpp
+	Static.cpp
+	TextOutput.cpp
+)
+
+target_include_directories(libhwbinder PUBLIC include ../system_core/include/ ${ANDROID_HEADERS_INCLUDE_DIRS})
+target_link_libraries(libhwbinder libutils libcutils libbinderthreadstate pthread)
+
+set_target_properties(libhwbinder PROPERTIES
+    VERSION ${PROJECT_VERSION_MAJOR}.${PROJECT_VERSION_MINOR}.${PROJECT_VERSION_PATCH}
+)
+
+install(TARGETS libhwbinder
+        LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+        PUBLIC_HEADER)
+
+install(DIRECTORY include/hwbinder DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/linux-hidl)
